Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,707 @@
+  change it into such normal form (see `WeierstrassCurve.toCharTwoJNeZeroNF`
+  and `WeierstrassCurve.toCharTwoJNeZeroNF_spec`).
+
+- `WeierstrassCurve.IsCharTwoNF` is the combination of the above two, that is, asserts that
+  a `WeierstrassCurve` is of form $Y^2 + XY = X^3 + a_2X^2 + a_6$ or
+  $Y^2 + a_3Y = X^3 + a_4X + a_6$. It is the normal form of characteristic = 2.
+
+  If the field is of characteristic = 2, then for any `WeierstrassCurve` there exists a change of
+  variables which will change it into such normal form
+  (`WeierstrassCurve.exists_variableChange_isCharTwoNF`).
+  See also `WeierstrassCurve.toCharTwoNF` and `WeierstrassCurve.toCharTwoNF_spec`.
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, normal form
+
+-/
+
+variable {R : Type*} [CommRing R] (W : WeierstrassCurve R)
+variable {F : Type*} [Field F] (E : EllipticCurve F)
+
+namespace WeierstrassCurve
+
+/-! ### Normal forms of characteristic ≠ 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2, if its $a_1, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_2X^2 + a_4X + a_6$. -/
+@[mk_iff]
+class IsCharNeTwoNF : Prop where
+  a₁ : W.a₁ = 0
+  a₃ : W.a₃ = 0
+
+section Quantity
+
+variable [W.IsCharNeTwoNF]
+
+@[simp]
+theorem a₁_of_isCharNeTwoNF : W.a₁ = 0 := IsCharNeTwoNF.a₁
+
+@[simp]
+theorem a₃_of_isCharNeTwoNF : W.a₃ = 0 := IsCharNeTwoNF.a₃
+
+@[simp]
+theorem b₂_of_isCharNeTwoNF : W.b₂ = 4 * W.a₂ := by
+  rw [b₂, a₁_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem b₄_of_isCharNeTwoNF : W.b₄ = 2 * W.a₄ := by
+  rw [b₄, a₃_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem b₆_of_isCharNeTwoNF : W.b₆ = 4 * W.a₆ := by
+  rw [b₆, a₃_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem b₈_of_isCharNeTwoNF : W.b₈ = 4 * W.a₂ * W.a₆ - W.a₄ ^ 2 := by
+  rw [b₈, a₁_of_isCharNeTwoNF, a₃_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem c₄_of_isCharNeTwoNF : W.c₄ = 16 * W.a₂ ^ 2 - 48 * W.a₄ := by
+  rw [c₄, b₂_of_isCharNeTwoNF, b₄_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem c₆_of_isCharNeTwoNF : W.c₆ = -64 * W.a₂ ^ 3 + 288 * W.a₂ * W.a₄ - 864 * W.a₆ := by
+  rw [c₆, b₂_of_isCharNeTwoNF, b₄_of_isCharNeTwoNF, b₆_of_isCharNeTwoNF]
+  ring1
+
+@[simp]
+theorem Δ_of_isCharNeTwoNF : W.Δ = -64 * W.a₂ ^ 3 * W.a₆ + 16 * W.a₂ ^ 2 * W.a₄ ^ 2 - 64 * W.a₄ ^ 3
+    - 432 * W.a₆ ^ 2 + 288 * W.a₂ * W.a₄ * W.a₆ := by
+  rw [Δ, b₂_of_isCharNeTwoNF, b₄_of_isCharNeTwoNF, b₆_of_isCharNeTwoNF, b₈_of_isCharNeTwoNF]
+  ring1
+
+end Quantity
+
+section VariableChange
+
+variable [Invertible (2 : R)]
+
+/-- There is an explicit change of variables of a `WeierstrassCurve` to
+a normal form of characteristic ≠ 2, provided that 2 is invertible in the ring. -/
+@[simps]
+def toCharNeTwoNF : VariableChange R := ⟨1, 0, ⅟2 * -W.a₁, ⅟2 * -W.a₃⟩
+
+instance toCharNeTwoNF_spec : (W.variableChange W.toCharNeTwoNF).IsCharNeTwoNF := by
```

Comment:
Oops, one last thing: what happens if you don't give them names? Are the self-generated names terrible?

Reply:
I haven't checked yet, but I prefer to give them explicit names; that's not harmful, right?